server/aesdsocket.c

```diff
@@ -85,126 +206,91 @@ syslog(LOG_USER, "Binded Connection from 9000");
 	redirect to STDIN, STDOUT, and STDERR
 */
-if(argc > 1){
-	if(strcmp(argv[1],"-d") == 0){
-		pid_t pid = fork();	//Create a child process using fork
-		if(pid < 0){
-			perror("Child Process not created: Daemon Process failed to create in step 1");
-			exit(-1);
-		}
-		else if(pid == 0){
-			if(setsid() < 0){
-				perror("Unable to set Id");
+	if(argc > 1){
+		if(strcmp(argv[1],"-d") == 0){
+			pid_t pid = fork();	//Create a child process using fork
+			if(pid < 0){
+				perror("Child Process not created: Daemon Process failed to create in step 1");
 				exit(-1);
 			}
-			
-			if(chdir("/") == -1){
-				perror("Unable to change directory");
-				exit(-1);
+			else if(pid == 0){
+				if(setsid() < 0){
+					perror("Unable to set Id");
+					exit(-1);
+				}
+				
+				if(chdir("/") == -1){
+					perror("Unable to change directory");
+					exit(-1);
+				}
+				open("/dev/null",O_RDWR);
+				dup(0);
+				dup(0);
+				syslog(LOG_USER,"Daemon Created Successfully");
+			}
+			else{
+				exit(0);
 			}
-			open("/dev/null",O_RDWR);
-			dup(0);
-			dup(0);
-			syslog(LOG_USER,"Daemon Created Successfully");
-		}
-		else{
-			exit(0);
 		}
 	}
-}
+
 /***************************************************************************************** Listening to the Server ***********************************************************************************/
+fd = open("/var/tmp/aesdsocketdata",O_RDWR|O_CREAT|O_APPEND, 0777);
+if(fd == -1){
+	perror("Unable to open the file");
+}
+syslog(LOG_USER,"File successfully opened");
 int listener = listen(sockfd, 10);
 if(listener < 0){											// Backlog value set as 10 as prescribed in the Beej's user guide on sockets.
 	perror("Failed to execute in the listening process");
 	return -1;
 }
 syslog(LOG_USER, "Listening phase completed");
-fd = open("/var/tmp/aesdsocketdata",O_RDWR|O_CREAT|O_APPEND, 0777);
-if(fd == -1){
-	perror("Unable to open the file");
-}
-syslog(LOG_USER,"File successfully opened");
-/*************************************************************************************************** Accepting packets *******************************************************************************/
-int bytes_received;
-int buffer_size = BUFFER_SIZE;
-struct sockaddr_in addr_client;
-socklen_t length;
-int newfd;
-int total_bytes_written_so_far = 0;
-while(1){
-	length = sizeof(addr_client);
-	newfd = accept(sockfd, (struct sockaddr *)&addr_client,&length);
-	if(newfd == -1){
-		perror("Unable to Accept connection from 9000");
-		exit(-1);
+/***************************************************************************************** Accepting the Packets *************************************************************************************/
+socklen_t clientfd;
+struct sockaddr_in clientadd;
+bool alarm_flag = false;
+pthread_mutex_init(&mutex_lock, NULL);
+TAILQ_INIT(&head);
+	while (1) {
+	// Spin a new thread on every connection accept
+		thread_data_t *datap = (thread_data_t *) malloc(sizeof(thread_data_t));
```

Comment:
@srki3050 a TERM signal received between this line and line 271 below will explain the valgrind memory leak you are seeing, because you will not free the memory allocated here since it's not yet in your linked list.